src/file-events/cpp/win_fsnotifier.cpp

```diff
@@ -70,28 +74,46 @@ void WatchPoint::listen() {
     }
 }
 
-void WatchPoint::handleEventsInBuffer(DWORD bytesTransferred) {
-    if (bytesTransferred == 0) {
-        // Got a buffer overflow => current changes lost => send INVALIDATE on root
-        log_info(server->getThreadEnv(), "Detected overflow for %ls", path.c_str());
-        server->reportEvent(FILE_EVENT_INVALIDATE, path);
-    } else {
-        int index = 0;
-        for (;;) {
-            FILE_NOTIFY_INFORMATION* current = (FILE_NOTIFY_INFORMATION*) &buffer[index];
-            handleEvent(current);
-            if (current->NextEntryOffset == 0) {
+void WatchPoint::handleEventsInBuffer(DWORD errorCode, DWORD bytesTransferred) {
+    server->handleEvents(this, errorCode, buffer, bytesTransferred);
+}
+
+void Server::handleEvents(WatchPoint* watchPoint, DWORD errorCode, const vector<BYTE>& buffer, DWORD bytesTransferred) {
+    JNIEnv* env = getThreadEnv();
+    const u16string& path = watchPoint->path;
+
+    try {
+        switch (errorCode) {
+            case ERROR_SUCCESS:
                 break;
+            case ERROR_OPERATION_ABORTED:
+                log_fine(env, "Finished watching '%s'", utf16ToUtf8String(path).c_str());
+                reportFinished(path);
+                return;
```

Comment:
```suggestion
                break;
```

Why not use `break` here as well?

Reply:
Because this is the point where we want to stop listening to this watch point. Using `break` here would result in the watcher being re-established.

Reply:
Ah, now I got it. For some reason I was under the impression that the `switch` statement is the last statement in the block. Maybe we could extract the rest of the code here into a separate method and call it from the `case ERROR_SUCCESS` branch? Then `break` would work fine.

Reply:
Yes, but that would add another variation to this method, maybe `handleEventsWithoutErrorCode()`. The extra indirection is not making the code more readable for me.

This `switch` at the top simply sorts out the cases we can handle early, and lets `ERROR_SUCCESS` pass through.

If you feel strongly I can try rearranging things a bit, but I don't think the end result will be any more readable.

Reply:
We can leave it like it is.